### PR TITLE
cli run-ios on device

### PR DESCRIPTION
**Pull request by LearningDave**
## **Summary**
At the moment the run-ios command from the react-native cli does only work for simulators. 
The pull request adds a new option to the existing command: **"--device 'device-name'" which installs and launches an iOS application on a connected device.**
This makes it easier to build a test environment using react-native for connected devices.

## **Tests**:
I've tested my code with the following commands:
react-native run-ios --device "Not existing device"
react-native run-ios --device
react-native run-ios --device "name-of-a-simulator"
react-native run-ios --device "name-of-connected-device" 

Output of the first three commands:
example_error_output

Additional to the manual command tests i've added a test file 'parseIOSDevicesList-test.js'.

## **Requirements**
I used **ios-deploy** In order to launch and install the .app-bundle on a connected device.
ios-deploy on github: https://github.com/phonegap/ios-deploy

EDIT:
Adding ios-deploy as dependency throws errors on non-osx machines. That's why i decided to
add ios-deploy as requirement instead of dependency.

### CODE REVIEW

**Comment by facebook-github-bot-3** [comment]
By analyzing the blame information on this pull request, we identified @grabbou and @tomkur to be potential reviewers.

**Comment by facebook-github-bot-3** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-3** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by facebook-github-bot-9** [comment]
@LearningDave updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/7429b5744760f59b80417be5539c0bffda2e1c67..a13cf370dd332c8cdcc5f5b8c8cbeb4dd9451d1e)

**Comment by mroswald** [comment]
cc @grabbou @frantic this PR adds a huge improvement for testing on real devices - please have a look

File: local-cli/runIOS/runIOS.js [hunk 1]
```diff
@@ -13,8 +13,10 @@ const fs = require('fs');
 const path = require('path');
 const findXcodeProject = require('./findXcodeProject');
 const parseIOSSimulatorsList = require('./parseIOSSimulatorsList');
+const parseIOSDevicesList = require('./parseIOSDevicesList');
 
 function runIOS(argv, config, args) {
+
```
Comment (mroswald) [inline local-cli/runIOS/runIOS.js]: please remove this empty line

File: local-cli/runIOS/runIOS.js [hunk 2]
```diff
@@ -103,5 +152,8 @@ module.exports = {
     description: 'Path relative to project root where the Xcode project '
      + '(.xcodeproj) lives. The default is \'ios\'.',
     default: 'ios',
-  }]
+  }, {
+    command: '--device [string]',
+    description: 'Explicitly set device to use',
+    }]
```
Comment (mroswald) [inline local-cli/runIOS/runIOS.js]: indentation seems wrong

File: local-cli/runIOS/runIOS.js [hunk 3]
```diff
@@ -90,6 +135,10 @@ module.exports = {
       desc: 'Pass a non-standard location of iOS directory',
       cmd: 'react-native run-ios --project-path "./app/ios"',
     },
+    {
+      desc: 'Run on a connected device, e.g. Maxs iPhone',
+      cmd: 'react-native run-ios --device "Max iPhone"',
```
Comment (mroswald) [inline local-cli/runIOS/runIOS.js]: please be consistent: "Max's iPhone" in both cases

**Comment by facebook-github-bot-4** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/a13cf370dd332c8cdcc5f5b8c8cbeb4dd9451d1e..e8703db7ae5df96d5c28198d993e13670a60cd5b)

**Comment by facebook-github-bot-3** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/e8703db7ae5df96d5c28198d993e13670a60cd5b..91518b3e92d53726daddab9237771f66af6633ab)

**Comment by facebook-github-bot-3** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/e8703db7ae5df96d5c28198d993e13670a60cd5b..91518b3e92d53726daddab9237771f66af6633ab)

File: local-cli/runIOS/runIOS.js [hunk 4]
```diff
@@ -25,15 +27,37 @@ function runIOS(argv, config, args) {
   const scheme = args.scheme || inferredSchemeName;
   console.log(`Found Xcode ${xcodeProject.isWorkspace ? 'workspace' : 'project'} ${xcodeProject.name}`);
 
+  if (args.device) {
+    const devices = parseIOSDevicesList(
+      child_process.execFileSync('xcrun', ['instruments', '-s'], {encoding: 'utf8'})
```
Comment (grabbou) [inline local-cli/runIOS/runIOS.js]: I am pretty sure I've seen similar PR in the past and there was discussion that you can pass `json` flag?

Comment (LearningDave) [inline local-cli/runIOS/runIOS.js]: Hi grabbou, 
yea sadly this only applies to `xcrun simctl list devices --json`,  which only lists simulators. 
I could implement it for the parseIOSSimulatorsList function. However the command i'm using to receive the real devices is `xcrun instruments -s`, which does not have a `json` flag as far as i know.

**Comment by facebook-github-bot-2** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/91518b3e92d53726daddab9237771f66af6633ab..c8dd87431e94c6d0ecf237e0d8d21199eafae1ad)

**Comment by facebook-github-bot-2** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/c8dd87431e94c6d0ecf237e0d8d21199eafae1ad..322f20faf66ac4968f33c32ba2866f882d815ba9)

File: local-cli/runIOS/runIOS.js (hunk 4 above)
Comment (LearningDave) [inline local-cli/runIOS/runIOS.js]: I'm using the `json` tag in order to receive the simulators now.

File: local-cli/runIOS/runIOS.js [hunk 5]
```diff
@@ -65,18 +87,46 @@ function runIOS(argv, config, args) {
...
-      return simulators[i];
+function runOnDevice(selectedDevice, scheme, xcodeProject){
+  buildProject(xcodeProject, selectedDevice.udid, scheme);
+  const iosDeployInstallArgs = [
+    '--bundle', 'build/Build/Products/Debug-iphoneos/' + scheme + '.app',
+    '--id' , selectedDevice.udid,
+    '--justlaunch'
+  ];
+  console.log('installing and launching app on connected device...');
+  child_process.spawnSync('ios-deploy', iosDeployInstallArgs, {stdio: 'inherit'});
```
Comment (javache) [inline local-cli/runIOS/runIOS.js]: Where does this binary come from? It's not a dependency of react-native as far as I can tell.

**Comment by facebook-github-bot-9** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/322f20faf66ac4968f33c32ba2866f882d815ba9..3b6edceb07edc07a5a3cacfcb709687954766b4b)

**Comment by facebook-github-bot-9** [comment]
@LearningDave updated the pull request.

File: local-cli/runIOS/runIOS.js (hunk 5 above)
Comment (LearningDave) [inline local-cli/runIOS/runIOS.js]: My bad. Added ios-deploy as dev-dependency.

**Comment by facebook-github-bot-9** [comment]
@LearningDave updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/06c746d51a68d67b985c90c44a455c69fa506171..6773a9069b9f63cd6f92c5440f2d20211e1448b7)

**Comment by facebook-github-bot-3** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/06c746d51a68d67b985c90c44a455c69fa506171..6773a9069b9f63cd6f92c5440f2d20211e1448b7)

**Comment by facebook-github-bot-6** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/6773a9069b9f63cd6f92c5440f2d20211e1448b7..b4786bf3efcf13023216bc956855157ec7d260c4)

**Comment by facebook-github-bot-0** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/b4786bf3efcf13023216bc956855157ec7d260c4..65d736ee7312cfc918d8e603081b46fd4be405aa)

File: local-cli/runIOS/runIOS.js (hunk 5 above)
Comment (javache) [inline local-cli/runIOS/runIOS.js]: What's the developer experience for this. Will people need to remember to also install ios-deploy?

Comment (mroswald) [inline local-cli/runIOS/runIOS.js]: Maybe add a check if `ios-deploy` is installed. If not, suggest to install it. Even better would be to add it to the dependencies, but as we have seen that throws errors on non-osx machines

**Comment by facebook-github-bot-5** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/65d736ee7312cfc918d8e603081b46fd4be405aa..c2f6e127a41ae39cf8ad3d3ad7fd4479b07a97f5)

File: local-cli/runIOS/runIOS.js (hunk 5 above)
Comment (LearningDave) [inline local-cli/runIOS/runIOS.js]: Good idea, i've added a message if ios-deploy is not installed globally:
failed
succeeded

**Comment by facebook-github-bot-3** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/c2f6e127a41ae39cf8ad3d3ad7fd4479b07a97f5..c26ac003a5037673fb6b66fbc07c00b5a6a82041)

File: local-cli/runIOS/runIOS.js [hunk 6]
```diff
@@ -24,32 +25,68 @@ function runIOS(argv, config, args) {
...
+      if (devices){
+        console.log('Could not find device with the name: "' + args.device + '".');
+        console.log('Choose one of the following:');
+        printFoundDevices(devices);
+      } else {
+        console.log('No iOS devices connected.');
+      }
+    }
+  } else if (args.udid) {
+      runOnDeviceByUdid(args.udid, scheme, xcodeProject, devices);
```
Comment (javache) [inline local-cli/runIOS/runIOS.js]: nit: indent with 2 spaces

**Comment by facebook-github-bot-9** [comment]
@LearningDave updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9414/files/c26ac003a5037673fb6b66fbc07c00b5a6a82041..89b072c66d36eb4f9d0439eda22c6affa465bdee)

**Comment by javache** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-3** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3821638) to review internal test results.

**Pull request closed without merging**